zerocopy-derive/tests/struct_try_from_bytes.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2023 The Fuchsia Authors
+//
+// Licensed under a BSD-style license <LICENSE-BSD>, Apache License, Version 2.0
+// <LICENSE-APACHE or https://www.apache.org/licenses/LICENSE-2.0>, or the MIT
+// license <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your option.
+// This file may not be copied, modified, or distributed except according to
+// those terms.
+
+#![allow(warnings)]
+
+mod util;
+
+use std::{marker::PhantomData, option::IntoIter};
+
+use {
+    static_assertions::assert_impl_all,
+    zerocopy::{FromBytes, FromZeros, TryFromBytes},
+};
+
+use crate::util::AU16;
+
+// A struct is `TryFromBytes` if:
+// - all fields are `TryFromBytes`
+
+#[derive(TryFromBytes, FromZeros, FromBytes)]
+struct Zst;
+
+assert_impl_all!(Zst: TryFromBytes);
+
+#[test]
+fn zst() {
+    // TODO(#5): Use `try_transmute` in this test once it's available.
+    let candidate = zerocopy::Ptr::from(&Zst);
+    // SAFETY: `candidate` is derived from the tested type, so it will contain
+    // `UnsafeCell`s at the same offsets as the tested type.
+    let is_bit_valid = unsafe { Zst::is_bit_valid(candidate) };
+    assert!(is_bit_valid);
+}
+
+#[derive(TryFromBytes, FromZeros, FromBytes)]
+struct One {
+    a: u8,
+}
+
+assert_impl_all!(One: TryFromBytes);
+
+#[test]
+fn one() {
+    // TODO(#5): Use `try_transmute` in this test once it's available.
+    let candidate = zerocopy::Ptr::from(&One { a: 42 });
+    // SAFETY: `candidate` is derived from the tested type, so it will contain
+    // `UnsafeCell`s at the same offsets as the tested type.
+    let is_bit_valid = unsafe { One::is_bit_valid(candidate) };
+    assert!(is_bit_valid);
+}
+
+#[derive(TryFromBytes, FromZeros)]
+struct Two {
+    a: bool,
+    b: Zst,
+}
+
+assert_impl_all!(Two: TryFromBytes);
+
+#[test]
+fn two() {
+    // TODO(#5): Use `try_transmute` in this test once it's available.
+    let candidate = zerocopy::Ptr::from(&Two { a: false, b: Zst });
+    // SAFETY: `candidate` is derived from the tested type, so it will contain
+    // `UnsafeCell`s at the same offsets as the tested type.
+    let is_bit_valid = unsafe { Two::is_bit_valid(candidate) };
+    assert!(is_bit_valid);
+}
+
+#[test]
+fn two_bad() {
+    // TODO(#5): Use `try_transmute` in this test once it's available.
+    let candidate = zerocopy::Ptr::from(&[2][..]);
+    // SAFETY:
+    // - The cast `cast(p)` is implemented exactly as follows: `|p: *mut T| p as
+    //   *mut U`.
+    // - The size of the object referenced by the resulting pointer is equal to
+    //   the size of the object referenced by `self`.
+    // - The alignment of `Unsized` is equal to the alignment of `[u8]`.
+    let candidate = unsafe { candidate.cast_unsized(|p| p as *mut Two) };
```

Comment:
Lol, I genuinely forgot `try_from_ref` existed. Actually, though, I think I prefer the current approach, since it allows us to focus the test on `is_bit_valid` (which is the item actually generated by `derive(TryFromBytes)`) and not take a dependency on `KnownLayout`.